### PR TITLE
Re-thought collections

### PR DESCRIPTION
**Pull request by XedinUnknown**
Aims to incorporate solution for #8.

### CODE REVIEW

File: src/CountableListInterface.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+use Traversable;
+use Countable;
+
+/**
+ * A list that can be counted.
+ *
+ * @since [*next-version*]
+ */
+interface CountableListInterface extends
```
Comment (mecha) [inline src/CountableListInterface.php:13]: From the looks of it, this is the "base" interface in this package. But I can't see how or why anything would require traversing _and_ counting. I was expecting `Traversable` to be the base interface for everything here, tbh.

File: src/MapInterface.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+namespace Dhii\Collection;
+
+use Psr\Container\ContainerInterface;
+
+/**
+ * A readable set.
+ *
+ * @since [*next-version*]
+ */
+interface MapInterface extends
```
Comment (mecha) [inline src/MapInterface.php:12]: Read my comments about `SetInterface` first.

Given my opinion on sets, my opinion here is that a map should be a `ContainerInterface` and a `Traversable`.

File: src/SetInterface.php [hunk 3]
```diff
@@ -2,82 +2,18 @@
...
+use Dhii\Data\Container\HasCapableInterface;
+
 /**
- * Something that can behave as a set with complete functionality.
+ * A list that can be checked for a key.
  *
  * @since [*next-version*]
  */
-interface SetInterface extends \Countable
+interface SetInterface extends
```
Comment (mecha) [inline src/SetInterface.php:13]: Looks very conceptually incorrect to me.

Here, it's declared as a traversable object which can be queried for a particular key's existence. I can't see how a consumer would require calling `has()` _and_ traversing. It seems to me like consumers would either need one or the other, in which case they ought to depend on `Traversable` or `HasCapableInterface` instead.

**Review by mecha — COMMENTED** [review]
To be honest, I find the interfaces quite confusing, both as concepts and as consumables. I've left comments on all the applicable files, as well as a conclusion below.

The intended reading order is (1) the comments (2) the conclusion.

----

We've been working without collections for quite a long time now. And we've managed, by depending on `Traversable`, `Countable`, `ContainerInterface` and other interfaces were needed, depending on the requirements.

In my opinion, this package should provide interfaces that allow us to do what we've been doing, and more. Such as depend on a `Traversable` `ContainerInterface`, a.k.a. `MapInterface`.

In all honesty, I'm suddenly not seeing the point of having interfaces for collections and sets.

**Comment by XedinUnknown** [comment]
Wow, this came as a surprize to me.

> But I can't see how or why anything would require traversing and counting

Something that wants to count would depend on `Countable`. Something that requires traversing would depend on `Traversable`. Countable provides a way to retrieve the count, which is not necessarily done by traversing. There's also a helper method somewhere which uses iteration as the last resort for counting. This is very important and intuitive.

> From the looks of it, this (`CoutableListInterface`) is the "base" interface in this package

It's one of the interfaces. Not everything requires counting, because not everything can be counted without traversing.

> Given my opinion on sets, my opinion here is that a map should be a `ContainerInterface` and a `Traversable`.

It is both `ContainerInterface` and `Traversable`. It extends `SetInterface` which extends `Traversable`. So, by extension, a Map is a readable set, and a set is a traversable that can be checked for key.

> I can't see how a consumer would require calling `has()` _and_ traversing. It seems to me like consumers would either need one or the other, in which case they ought to depend on` Traversable` or `HasCapableInterface` instead

But I can. `has()` is useful for testing for a particular key's existence, which is not necessarily done by expensive iteration. `isset()` and `array_key_exists()` work that way: they do not look at every element, but just check.

> We've been working without collections for quite a long time now. And we've managed, by depending on Traversable, Countable, ContainerInterface and other interfaces were needed, depending on the requirements.

It has been established that something that is a traversable container could be very useful - because you may want to copy it. This is effectively what a map achieves. But a map groups a lot of different functionality, and not all of it will always be useful or feasible to implement. What I did here was segregate each combination of potentially useful functionality into an interface of its own. Then the names came to me. A multitude that can be tested for whether it contains something is a Set - which is what the base Collection used to be. A set that can have a specific item retrieved by key is a readable Map - or an iterable container, depending how you look at it. Not all such containers need to be counted, or in fact can be counted correctly - such as a DI container. But they still can be iterable - perhaps recursively. And thus we have a Map and CountableMap separately. This makes a lot of sense to me. I encourage you to review this again, with this explanation in mind.

**Comment by mecha** [comment]
> Countable provides a way to retrieve the count, which is not necessarily done by traversing. There's also a helper method somewhere which uses iteration as the last resort for counting. This is very important and intuitive.

Doesn't sound like a argument in favor of `CountableListInterface`. The helper method is, imo, irrelevant. Counting by traversing as a last resort is fine, as long as the subject is `Traversable`. It doesn't mean the method, or any other consumer, **depends** on something that is both. It's also wrong imo to depend on both in the helper method. Because a subject should be allowed _either_ countable or traversable. Being required to be both is a restriction that becomes a flaw.

I've yet to think of, or be presented, a reasonable use case for this interface.

> It is both ContainerInterface and Traversable. It extends SetInterface which extends Traversable. So, by extension, a Map is a readable set, and a set is a traversable that can be checked for key.

Yes, that's what it is ultimately. But I disagree with `SetInterface`.

> has() is useful for testing for a particular key's existence, which is not necessarily done by expensive iteration.

My main point is that it doesn't make sense to depend on something that is both. I may just be unable to see a need for it, but this is the way I see it:

If you're requiring `SetInterface`, then you're indirectly saying that you need something that can be iterated over and checked for key/element existence. In other words, you're implying that you're going to iterate and check for keys/elements. My argument is, why would a consumer need to do both?

If there are cases where this is required, then fine. I'm game. But I fail to see how or why.

> isset() and array_key_exists() work that way: they do not look at every element, but just check.

Excellent point. Consumers that require the equivalent of this can depend on `HasCapableInterface`.

>  It has been established that something that is a traversable container could be very useful ... But a map groups a lot of different functionality, and not all of it will always be useful or feasible to implement. What I did here was segregate each combination of potentially useful functionality into an interface of its own.

It was you who said that _"It is both ContainerInterface and Traversable. It extends SetInterface which extends Traversable. So, by extension, a Map is a readable set, and a set is a traversable that can be checked for key."_, so to me this sounds like a contradiction. It's still a traversable container.

> And thus we have a Map and CountableMap separately. This makes a lot of sense to me. I encourage you to review this again, with this explanation in mind.

But I fail to see the need for the latter. Think about it. If you're depending on having some "count" or "size" or "quantity", then isn't also depending on the subject being a "map" limiting? Sure, they're separate and that's great. But I need to know how the countable-versions will be used, if there is even a need for them. I encourage you to reconsider these interfaces again, this time keeping in mind what we have identified as a _need_, like we did with a traversable container.

**Comment by XedinUnknown** [comment]
> It doesn't mean the method, or any other consumer, depends on something that is both

But the consumer **may** need to depend on something that is both. At the same time, this is all that the consumer may need. Also, intuitively it feels right to be able to know the size of something that can be traversed.

> Being required to be both is a restriction that becomes a flaw.

Unless this is what the logic requires. Otherwise, the consumer is free to depend on either.

> Yes, that's what it is ultimately. But I disagree with `SetInterface`.

"I disagree" should be followed by the reason in order to make a point.

> If you're requiring `SetInterface`, then you're indirectly saying that you need something that can be iterated over and checked for key/element existence.

Directly.

> In other words, you're implying that you're going to iterate and check for keys/elements. My argument is, why would a consumer need to do both?

"Why would a consumer need to [insert requirement]" is besides the point. It's like asking "Why would somebody need to use our software, if we're only making it for ourselves?". It's besides the point. I only think of why the consumer may need to do something when I am asked to justify something that is right. I do it because it's right, not because there may be a need to do something in the future. But if you really want an explanation, then I can say that I have encountered cases where the consumer itself may only need a part of the interface, but then may need to forward the argument to something that needs the other part of the interface. An example where you may need to do both is when you only want to iterate over something if it has a particular value. Why a consumer may need to do this? I don't know. But it doesn't matter.

> Excellent point. Consumers that require the equivalent of this can depend on `HasCapableInterface`. 

See reply above.

> to me this sounds like a contradiction. It's still a traversable container.

Yes it is. But a traversable container allows `has()` and `foreach()`, which means that a traversable container is also a set.

> If you're depending on having some "count" or "size" or "quantity", then isn't also depending on the subject being a "map" limiting?

If it is limiting, then maybe you don't need both, and then you should depend on an interface with the required subset of methods. If, on the other hand, the particular combination is required, then it isn't limiting.

> I encourage you to reconsider these interfaces again, this time keeping in mind what we have identified as a need, like we did with a traversable container.

Our needs are not the only needs there are. So, unless you give a reason for how these interfaces may be wrong, like not segregated enough or missing some natural methods, I will merge this.

I don't need a reason to do the right thing. I need a reason for why it may not be right. But this appears to be an architect VS logician debate, to me. YAGNI is not a principle I tend to follow.

**Comment by mecha** [comment]
To me it's not about following YAGNI, of which I'm not a fan either. But standards are built out of necessity. We identify that there is something that we need to achieve, design a flexible and interoperable way to do it, and we created the interfaces. This is what we've always done and it's worked great. Just look at validation, expressions, modules, etc.

Reading the contents of this PR made me realize that we don't yet have a _need_ for collections. All we've identified is a need for traversable container. I don't want us to start creating interfaces and standards based on "intuitively feeling right" with a "gut feeling that it _may_ be needed". Else we run the risk of creating bloated standards that might need drastic changes in the future that break backwards compatibility.

This something I learned from expressions. When I came to work with those interfaces, I realized just how unsuitable they were for my very generic and common purpose: traversal and interpretation. At both times of refactoring, they seemed to be intuitively correct until we (1) extracted `EvaluableInterface` and (2) later even made expressions non-evaluable and hierarchy-only. Even though an "expression" is intuitively something that is evaluable. That refactoring is something we could have avoided by simply identifying a _need_ first through an actual dependent.

My argument is not that these interfaces are wrong. Neither is it the rhetorical "why would a consumer need this?". It is the **literal** "why would a consumer need this?".

I recommend that until such a time when we truly identify a need for list/set/collection interfaces, and we _know_ what we want to do with them, we hold off creating them. So far, the only requirement is `MapInterface`, and we already know that we want it to extend `Traversable` and `ContainerInterface`. It doesn't matter if in the future it extends some other interfaces to be both of those things (like how it is right now through `SetInterface` and `ContainerInterface`), because we _know_ that it should still be both of those things, as per our identified requirements.

I encourage you to have a good hard look at whatever prompted you to write this package (I'm assuming that it's modules) and identify what it is exactly that the consumer code requires. You may have already done this, but I can't assume that - no scenarios were given as replies to my questions about "why a consumer would need". But I _can_ encourage.

Of course, if you disagree with this I may be inclined to approve this PR anyway. Time is a commodity we don't have, and I'd rather either one of us eventually turn out to be wrong about this standard, than waste time debating. Especially when my arguments are made out to not be those of an architect. In the end, these are my 2 cents based on experience. If you want to merge this anyway, please let me know and I'll put my trust in your and approve the PR.

**Comment by XedinUnknown** [comment]
The need identified is that of a traversable container. This, as you've pointed out, is the basic drive behind this package.
However, a traversable container is many things. It implements `HasCapableInterface`, which is made to follow ISP - since the standard PSR container does not have enough segregation. `MapInterface` is a requirement, but on its own it also doesn't provide enough segregation. ISP is a principle to follow, and it's not about what the consumer may need. Therefore, I segregated subsets into interfaces of their own. I do not concern myself with their usefulness - only with ISP. While doing that, I remembered those time when I needed something to be two things at the same time. Truly, if you need something to be both a traversable and a container, then how can you tell that something will not need to be both countable and traversable? I can see a potential need for being able to count elements in a container without iteration. And this prompts for segregation of a subset. That's all I did. I do not see how, given that we follow the same SOLID principles in consuming code, this kind of subset can negatively impact the quality of the code, or its maintainability.

The "logician VS architect" was a comparison of MBTI types, rather than positions, my friend.  ❤️

**Comment by mecha** [comment]
I _**was**_ going to just agree and approve. But that's just not me.

----

> ISP is a principle to follow, and it's not about what the consumer may need.

Then my entire life has been a lie. Because AFAIK, the most important reason for doing ISP is for the consumer to depend only on what they need and nothing more, while also allowing implementations to not be burdened. (*Edit*: this seems to be [the very definition of ISP][1], as a matter of fact).

It's one thing to do ISP, and another to go create combinations based on the gut feeling that they _might_ be needed. By all means, segregate. And by all means if you're convinced, or **even better** have other work that is waiting to depend on `CountableListInterface`, do it. Tell me. I'll approve. You'll merge.

> I do not concern myself with their usefulness - only with ISP.

Part of me cried a bit. Read previous reply.

> MapInterface is a requirement, but on its own it also doesn't provide enough segregation.

`MapInterface` extends `Dhii\Data\ContainerInterface` and `Traversable`.

`Dhii\Data\Container` is both `GetCapableInterface` and `HasCapableInterface`.

What's the problem? This looks pretty segregated to me.

> Truly, if you need something to be both a traversable and a container, then how can you tell that something will not need to be both countable and traversable? I can see a potential need for being able to count elements in a container without iteration.

Again. I agree. Counting without iterating: definitely a need. Depend on `Countable`.

I have yet to see something that requires this. I'm not blocking you from merging this in. I'm just concerned with bloating the standard just for the sake of having a concept be represented by an interface.

Although, a part of me wouldn't use such a combo interface even if I had a valid use case of counting and iterating. Because as you said, we have helpers and abstractions for this. An iterable is countable. But if an iterator implementation wants to be performant, it can just implement `Countable`.

```php
public function _foo(Traversable $t)
{
    $c = _count($t);
    // ...
}
public function _count($arg)
{
    if ($arg instanceof Countable) {
        return count($arg);
    }
    // ...
}
```

This is more flexible IMO. Not everything has to be supported at an interface level. Many things, in fact, ought to be supported at a concrete level. Think about `StringableInterface` for a moment.

> I do not see how, given that we follow the same SOLID principles in consuming code, this kind of subset can negatively impact the quality of the code, or its maintainability.

It's not about negative impacts or anything like that. It's about **bloated standards** without **real purpose**.

----

We've been down this road before; modelling something based on what we _think_ is most conceptually correct, only to later depend on it and find that it's not suitable and we go back to refactor.

* Expressions
* State Machine
* Invocables
* Config
* Trees
* _et. al_

I'm sorry if I misinterpreted your comment about logician vs architect. But given the sentence that preceded it, it sounded like you really believed like you are doing the right thing, and that _that_ is what it means to be an architect. That makes me feel like I've lost the argument before it even began.

[1]: https://en.wikipedia.org/wiki/Interface_segregation_principle

**Comment by XedinUnknown** [comment]
Again, I feel like you are leaving my most important point without a counter-argument.

In the scenario where you have one function iterate and the other count, this is OK because (1) objects that are `Traversable` are often `Countable`; (2) it is always possible to count something that is `Traversable` even if it isn't `Countable`. But like I said, this isn't always going to be the case.

Either way, it us useful for a container to be countable, because (1) it may increase performance; (2) counting a container by iterating (think composite container) may not yield usable results.

If something `extends ContainerInterface, HasCapableInterface, Traversable`, ISP tells me that there must be something that only `extends HasCapableInterface, Traversable`. Because otherwise, I am forcing the consumer that needs these two depend on something that has all 3 - which breaks ISP. In the beginning, it was called a `HasCapableListInterface`. This name is not very user-friendly. Then there was a derivative `HasCapableCountableListInterface` - for the same reasons of ISP. Then I realized that `HasCapableListInterface` is in fact a set. Then, `HasCapableCountableListInterface` is a countable set. Everything made sense to me at that point. This collection of interfaces makes it possible for a consumer to be comfortable and have everything and only what it needs - whatever it may need. That's my job. And this is a job well done. I would rather find out that some of these interfaces are extra, than have all consumers that need a countable set create their own `CountableSetInterface` and lose interop.

I also want to say that I super appreciate this:
> Time is a commodity we don't have, and I'd rather either one of us eventually turn out to be wrong about this standard, than waste time debating.

**Pull request merged**